indico/modules/events/persons/controllers.py

```diff
@@ -63,62 +76,84 @@ def get_persons(self):
-
+            if registration:
+                data['registrations'].append(registration)
             data['person'] = event_person
-            data['roles']['chairperson'] = event_person in chairpersons
-
-            for person_link in event_person.abstract_links:
-                if person_link.abstract.is_deleted:
-                    continue
-
-                url = url_for('abstracts.display_abstract', person_link.abstract, management=True)
-                data['abstracts'][person_link.abstract_id] = {'title': person_link.abstract.verbose_title, 'url': url}
-                data['roles']['author'] = True
-
-            for person_link in event_person.session_block_links:
-                if person_link.session_block.session.is_deleted:
-                    continue
+            if event_person in chairpersons:
+                data['roles']['chairperson'] = BUILTIN_ROLES['chairperson'].copy()
 
-                data['session_blocks'][person_link.session_block_id] = {'title': person_link.session_block.full_title}
-                data['roles']['convener'] = True
-
-            for person_link in event_person.contribution_links:
-                if not person_link.is_speaker:
-                    continue
-                contrib = person_link.contribution
-                if contrib.is_deleted:
-                    continue
-
-                url = url_for('contributions.manage_contributions', self.event, selected=contrib.friendly_id)
-                data['contributions'][contrib.id] = {'title': contrib.title, 'url': url}
-                data['roles']['speaker'] = True
-
-            for person_link in event_person.subcontribution_links:
-                subcontrib = person_link.subcontribution
-                contrib = subcontrib.contribution
-                if subcontrib.is_deleted or contrib.is_deleted:
-                    continue
+            if self.event.type == 'lecture':
+                continue
 
-                url = url_for('contributions.manage_contributions', self.event, selected=contrib.friendly_id)
-                data['subcontributions'][subcontrib.id] = {'title': '{} ({})'.format(contrib.title, subcontrib.title),
-                                                           'url': url}
-                data['roles']['speaker'] = True
+            if self.event.has_feature('abstracts'):
+                abstracts = {
+                    person_link.abstract_id: {
+                        'title': person_link.abstract.verbose_title,
+                        'url': url_for('abstracts.display_abstract', person_link.abstract, management=True)
+                    }
+                    for person_link in event_person.abstract_links if not person_link.abstract.is_deleted
+                }
+
+                if abstracts:
+                    data['roles']['author'] = BUILTIN_ROLES['author'].copy()
+                    data['roles']['author']['elements'] = abstracts
+
+            session_blocks = {
```

Comment:
personally I find it a bit unreadable, what about a normal `for` statement instead of the list comprehensions

Reply:
agreed, this comprehension is pretty ugly. but i don't think using a normal for loop will be much better. why not add some helper functions that generate the data dicts? then the comprehension would be much more readable.

Reply:
Also, I pair-programmed this together with @pferreir, we thought about the possibility of using a helper function but, if you notice, there are some differences among each role. For instance, `abstracts` has `title` and `url` whereas `session_blocks` just has `title`, for the `subcontributions` we need to access the contribution title so the logic is different from the `contributions` itself and so on. I think the helper function would be very complex, unless you have a better idea on how to implement it.

Reply:
i meant having one helper function for each type. writing one big (and ugly) function would be somewhat ugly indeed